File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/workflow.py
from __future__ import annotations

import copy
import os
import uuid
from pathlib import Path

from .common import artifact_paths, read_json, utc_now, write_json
from .contracts import build_delegate_prompt, build_system_prompt, default_completion_contract
from .delegate import base_envelope, write_failure_envelope
from .job_state import finalize_job_state, initialize_job, update_job
from .pipeline import execute_request_pipeline
from .request import finalize_request


def _step_summary(envelope: dict, role: str, job_path: str) -> dict:
    completion = envelope.get("completion", {})
    return {
        "role": role,
        "job_path": job_path,
        "ok": envelope.get("ok", False),
        "session_id": envelope.get("session_id"),
        "summary": completion.get("summary") or envelope.get("result") or "",
        "error_type": envelope.get("error_type"),
        "boundary_status": (envelope.get("boundary") or {}).get("status"),
        "findings": envelope.get("findings", []),
        "open_risks": envelope.get("open_risks", []),
        "changed_files": envelope.get("changed_files", []),
        "diff_summary": envelope.get("diff_summary", []),
        "files_examined": envelope.get("files_examined", []),
        "test_commands": envelope.get("test_commands", []),
        "suggested_actions": envelope.get("suggested_actions", []),
        "verification_status": (envelope.get("verification") or {}).get("status"),
    }


def _dedupe_strings(values: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for value in values:
        if not value or value in seen:
            continue
        seen.add(value)
        result.append(value)
    return result


# Inheritance policy: explicit specification of workflow child request inheritance
# Defines which fields from parent task_packet are inherited by child roles
EXECUTION_POLICY_INHERITANCE_SPEC = {
    # Fields inherited by all roles if explicitly set in parent (preserving exact values)
    "all_roles": ["command_allowlist", "exclude_globs", "max_budget_usd", "max_changed_files", "max_turns", "observe_roots"],
    # Fields inherited only by implementer role
    "implementer_only": ["allowed_write_paths"],
}

VERIFICATION_CONTRACT_INHERITANCE_SPEC = {
    # Fields inherited by all roles if explicitly set in parent
    "all_roles": ["commands"],
    # Fields inherited only by implementer role
    "implementer_only": ["auto", "fail_on_error"],
    # Fields inherited by non-implementer roles (critic, explorer)
    "non_implementer_only": ["fail_on_error"],
}


def _inherit_child_task_packet(parent_request: dict, role: str) -> dict:
    """Inherit child task packet from parent using explicit inheritance specs.

    Preserves exact values (including empty lists, false, and zero) for fields
    specified in inheritance specs. Absent fields in parent do not add inherited fields.
    """
    packet = copy.deepcopy(parent_request["task_packet"])
    policy = copy.deepcopy(packet.pop("execution_policy", {}))
    verification = copy.deepcopy(packet.pop("verification_contract", {}))
    packet.pop("assistant_role", None)
    packet.pop("workflow_roles", None)

    inherited_policy: dict = {}
    # Inherit execution_policy fields that are explicitly present in parent,
    # preserving exact values including empty lists and false/zero values.
    for key in EXECUTION_POLICY_INHERITANCE_SPEC["all_roles"]:
        if key in policy:
            inherited_policy[key] = policy[key]
    # allowed_write_paths is implementer-specific.
    if role == "implementer":
        for key in EXECUTION_POLICY_INHERITANCE_SPEC["implementer_only"]:
            if key in policy:
                inherited_policy[key] = policy[key]
    if inherited_policy:
        packet["execution_policy"] = inherited_policy

    inherited_verification: dict = {}
    # commands apply to all roles if explicitly set in parent.
    for key in VERIFICATION_CONTRACT_INHERITANCE_SPEC["all_roles"]:
        if key in verification:
            inherited_verification[key] = verification[key]
    if role == "implementer":
        # auto and fail_on_error apply to implementer if explicitly set in parent.
        for key in VERIFICATION_CONTRACT_INHERITANCE_SPEC["implementer_only"]:
            if key in verification:
                inherited_verification[key] = verification[key]
    else:
        # fail_on_error applies to non-implementer roles if explicitly set in parent.
        for key in VERIFICATION_CONTRACT_INHERITANCE_SPEC["non_implementer_only"]:
            if key in verification:
                inherited_verification[key] = verification[key]
    if inherited_verification:
        packet["verification_contract"] = inherited_verification

    return packet


def _prepare_role_request(parent_request: dict, role: str, prior_steps: list[dict], parent_job_path: str) -> dict:
    child_request = copy.deepcopy(parent_request)
    child_request["assistant_role"] = role
    child_request["workflow_roles"] = [role]
    child_request["session_id"] = str(uuid.uuid4())
    child_request["resume_session_id"] = None
    child_request["completion_contract"] = default_completion_contract(role)
    child_request["task_packet"] = _inherit_child_task_packet(parent_request, role)
    child_request["lineage"] = {
        "action": "workflow_step",
        "parent_job_path": parent_job_path,
        "workflow_root_session_id": parent_request["session_id"],
    }
    child_request["skip_ledger"] = True
    child_request = finalize_request(child_request)
    child_request["system_prompt"] = build_system_prompt(parent_request["base_system_prompt"], role)
    child_request["prompt"] = build_delegate_prompt(
        child_request["task_packet"],
        assistant_role=role,
        completion_contract=child_request["completion_contract"],
        delta_prompt=child_request.get("delta_prompt"),
        prior_steps=prior_steps,
    )
    return child_request


def _append_step_job_path(parent_job_paths: list[str], job_path: str) -> list[str]:
    if job_path in parent_job_paths:
        return list(parent_job_paths)
    return [*parent_job_paths, job_path]


def _execute_child_step(
    child_request: dict,
    child_dir: Path,
    *,
    role: str,
    workflow_step: int,
    update_parent_job,
    parent_paths: dict[str, Path],
    parent_event_count: int,
) -> tuple[dict, int]:
    child_paths = artifact_paths(child_dir)
    child_paths["lock"].touch()
    write_json(child_paths["request"], child_request)
    initialize_job(child_paths, child_request)
    update_job(
        child_paths,
        state="running",
        started_at=utc_now(),
        pid=os.getpid(),
        current_role=role,
        last_error=None,
    )

    parent_job = read_json(parent_paths["job"]) or {}
    step_job_paths = _append_step_job_path(parent_job.get("step_job_paths", []), str(child_dir))
    update_parent_job(
        current_role=role,
        current_step_job_path=str(child_dir),
        step_job_paths=step_job_paths,
        workflow_step=workflow_step,
    )

    def on_spawn(delegate_pid: int) -> None:
        update_job(child_paths, delegate_pid=delegate_pid)
        update_parent_job(delegate_pid=delegate_pid, current_step_job_path=str(child_dir))

    def on_event(event: dict, event_count: int) -> None:
        nonlocal parent_event_count
        parent_event_count += 1
        timestamp = utc_now()
        update_job(
            child_paths,
            event_count=event_count,
            last_event_at=timestamp,
            last_event_type=event.get("type"),
        )
        update_parent_job(
            event_count=parent_event_count,
            last_event_at=timestamp,
            last_event_type=event.get("type"),
            current_step_job_path=str(child_dir),
        )

    try:
        envelope = execute_request_pipeline(
            child_request,
            child_dir,
            on_spawn=on_spawn,
            on_event=on_event,
        )
    except Exception as exc:
        write_failure_envelope(
            child_request,
            child_dir,
            "workflow_step_error",
            str(exc),
            exit_code=1,
        )
        update_job(
            child_paths,
            state="failed",
            finished_at=utc_now(),
            current_role=None,
            delegate_pid=None,
            last_error=str(exc),
        )
        update_parent_job(delegate_pid=None, current_step_job_path=None)
        raise

    update_job(
        child_paths,
        state=finalize_job_state(envelope),
        finished_at=utc_now(),
        assistant_role=envelope.get("assistant_role", child_request.get("assistant_role")),
        current_role=None,
        delegate_pid=None,
        last_error=envelope.get("error_message"),
        task_type=envelope.get("task_type") or child_request.get("task_type"),
        workflow_roles=envelope.get("workflow_roles", child_request.get("workflow_roles")),
        workflow_step=1,
        workflow_total_steps=1,
    )
    update_parent_job(delegate_pid=None, current_step_job_path=None)
    return envelope, parent_event_count


def _aggregate_boundary(step_envelopes: list[dict]) -> dict:
    violations: list[dict] = []
    for envelope in step_envelopes:
        violations.extend((envelope.get("boundary") or {}).get("violations", []))
    return {
        "status": "violated" if violations else "passed",
        "violations": violations,
    }


def _aggregate_verification(step_envelopes: list[dict]) -> dict:
    statuses = [(envelope.get("verification") or {}).get("status") for envelope in step_envelopes]
    commands: list[str] = []
    results: list[dict] = []
    fail_on_error = False

    for envelope in step_envelopes:
        verification = envelope.get("verification") or {}
        commands.extend(verification.get("commands", []))
        results.extend(verification.get("results", []))
        fail_on_error = fail_on_error or verification.get("fail_on_error", False)

    if any(status == "failed" for status in statuses):
        status = "failed"
    elif any(status == "passed" for status in statuses):
        status = "passed"
    else:
        status = "skipped"

    return {
        "status": status,
        "fail_on_error": fail_on_error,
        "commands": _dedupe_strings(commands),
        "results": results,
    }


def execute_workflow(
    request: dict,
    artifacts_dir: Path,
    *,
    update_parent_job,
) -> dict:
    paths = artifact_paths(artifacts_dir)
    workflow_envelope = base_envelope(paths, request)
    workflow_envelope["assistant_role"] = "supervisor"
    workflow_envelope["task_type"] = request["task_packet"]["task_type"]
    workflow_envelope["workflow_roles"] = request["workflow_roles"]
    workflow_envelope["lineage"] = request.get("lineage")
    workflow_envelope["completion_contract"] = {
        "name": request["completion_contract"]["name"],
        "role": request["completion_contract"]["role"],
    }

    step_artifacts = artifacts_dir / "steps"
    step_artifacts.mkdir(parents=True, exist_ok=True)
    workflow_envelope["artifacts"]["steps_dir"] = str(step_artifacts)
    prior_steps: list[dict] = []
    step_summaries: list[dict] = []
    step_envelopes: list[dict] = []
    parent_event_count = 0

    for index, role in enumerate(request["workflow_roles"], start=1):
        update_parent_job(
            current_role=role,
            workflow_step=index,
            workflow_total_steps=len(request["workflow_roles"]),
            completed_roles=[step["role"] for step in step_summaries],
        )
        child_dir = step_artifacts / f"{index:02d}-{role}"
        child_dir.mkdir(parents=True, exist_ok=True)
        child_request = _prepare_role_request(request, role, prior_steps, str(paths["artifacts_dir"]))
        child_envelope, parent_event_count = _execute_child_step(
            child_request,
            child_dir,
            role=role,
            workflow_step=index,
            update_parent_job=update_parent_job,
            parent_paths=paths,
            parent_event_count=parent_event_count,
        )
        step_envelopes.append(child_envelope)
        summary = _step_summary(child_envelope, role, str(child_dir))
        step_summaries.append(summary)
        prior_steps.append(summary)

        if role != "critic" and not child_envelope.get("ok", False):
            break

    final_child = step_envelopes[-1] if step_envelopes else {}
    all_findings = []
    critic_findings = []
    open_risks = []
    total_cost_usd = 0.0
    total_cost_seen = False
    duration_ms = 0
    duration_seen = False
    num_turns = 0
    turns_seen = False
    permission_denials: list[dict] = []
    tool_uses: list[dict] = []
    tool_use_count = 0

    for summary in step_summaries:
        all_findings.extend(summary.get("findings", []))
        if summary["role"] == "critic":
            critic_findings.extend(summary.get("findings", []))
        open_risks.extend(summary.get("open_risks", []))
    for envelope in step_envelopes:
        cost = envelope.get("total_cost_usd")
        if isinstance(cost, (int, float)):
            total_cost_seen = True
            total_cost_usd += float(cost)
        child_duration = envelope.get("duration_ms")
        if isinstance(child_duration, int):
            duration_seen = True
            duration_ms += child_duration
        child_turns = envelope.get("num_turns")
        if isinstance(child_turns, int):
            turns_seen = True
            num_turns += child_turns
        permission_denials.extend(envelope.get("permission_denials", []))
        tool_uses.extend(envelope.get("tool_uses", []))
        tool_use_count += int(envelope.get("tool_use_count") or len(envelope.get("tool_uses", [])))

    steps_failed = any(not envelope.get("ok", False) for envelope in step_envelopes)
    boundary = _aggregate_boundary(step_envelopes)
    verification = _aggregate_verification(step_envelopes)
    status = "failed" if steps_failed else "needs_review" if critic_findings else "completed"

    workflow_envelope["workflow"] = {
        "roles": request["workflow_roles"],
        "steps": step_summaries,
    }
    workflow_envelope["structured_output"] = {
        "status": status,
        "summary": final_child.get("completion", {}).get("summary") or final_child.get("result") or "workflow finished",
        "steps": [
            {
                "role": item["role"],
                "job_path": item["job_path"],
                "ok": item["ok"],
                "summary": item["summary"],
            }
            for item in step_summaries
        ],
        "open_risks": _dedupe_strings(open_risks),
    }
    workflow_envelope["completion"] = workflow_envelope["structured_output"]
    workflow_envelope["findings"] = all_findings
    workflow_envelope["open_risks"] = _dedupe_strings(open_risks)
    workflow_envelope["result"] = final_child.get("result")
    workflow_envelope["total_cost_usd"] = total_cost_usd if total_cost_seen else None
    workflow_envelope["duration_ms"] = duration_ms if duration_seen else None
    workflow_envelope["num_turns"] = num_turns if turns_seen else None
    workflow_envelope["permission_denials"] = permission_denials
    workflow_envelope["tool_use_count"] = tool_use_count
    workflow_envelope["tool_uses"] = tool_uses
    workflow_envelope["boundary"] = boundary
    workflow_envelope["verification"] = verification
    workflow_envelope["session_id"] = request["session_id"]
    workflow_envelope["ok"] = not steps_failed and not critic_findings

    if steps_failed:
        failed_step = next((item for item in step_envelopes if not item.get("ok", False)), {})
        workflow_envelope["error_type"] = failed_step.get("error_type") or "workflow_step_failed"
        workflow_envelope["error_message"] = failed_step.get("error_message") or "workflow step failed"
    elif critic_findings:
        workflow_envelope["ok"] = False
        workflow_envelope["error_type"] = "critic_findings"
        workflow_envelope["error_message"] = "critic found issues"

    update_parent_job(
        completed_roles=[step["role"] for step in step_summaries],
        current_role=None,
        workflow_step=len(step_summaries),
    )
    write_json(paths["normalized"], workflow_envelope)
    from .handoff import shape_handoff

    write_json(paths["handoff"], shape_handoff(workflow_envelope))
    return workflow_envelope

File: plugins/agent-skills/skills/claude-delegate/src/lib/workflow.ts
import { mkdirSync } from "node:fs";
import path from "node:path";

import { artifactPaths, readJson, utcNow, writeJson } from "./common.js";
import { dedupeStrings } from "./contracts.js";
import { baseEnvelope, writeFailureEnvelope } from "./delegate.js";
import { shapeHandoff } from "./handoff.js";
import {
  finalizeJobState,
  initializeJob,
  updateJob,
} from "./job-state.js";
import { executeRequestPipeline } from "./pipeline.js";
import {
  aggregateBoundary,
  aggregateVerification,
  stepSummary,
} from "./workflow-aggregation.js";
import { prepareRoleRequest } from "./workflow-inheritance.js";

function appendStepJobPath(
  parentJobPaths: string[],
  jobPath: string,
): string[] {
  if (parentJobPaths.includes(jobPath)) return [...parentJobPaths];
  return [...parentJobPaths, jobPath];
}

async function executeChildStep(
  childRequest: Record<string, unknown>,
  childDir: string,
  opts: {
    role: string;
    workflowStep: number;
    updateParentJob: (changes: Record<string, unknown>) => void;
    parentPaths: ReturnType<typeof artifactPaths>;
    parentEventCount: number;
  },
): Promise<[Record<string, unknown>, number]> {
  const childPaths = artifactPaths(childDir);
  // Ensure lock file exists
  const { writeFileSync, existsSync } = await import("node:fs");
  if (!existsSync(childPaths.lock)) {
    writeFileSync(childPaths.lock, "");
  }
  writeJson(childPaths.request, childRequest);
  initializeJob(childPaths, childRequest);
  updateJob(childPaths, {
    state: "running",
    started_at: utcNow(),
    pid: process.pid,
    current_role: opts.role,
    last_error: null,
  });

  const parentJob = readJson(opts.parentPaths.job) ?? {};
  const stepJobPaths = appendStepJobPath(
    (parentJob.step_job_paths as string[]) ?? [],
    childDir,
  );
  opts.updateParentJob({
    current_role: opts.role,
    current_step_job_path: childDir,
    step_job_paths: stepJobPaths,
    workflow_step: opts.workflowStep,
  });

  let parentEventCount = opts.parentEventCount;

  function onSpawn(delegatePid: number): void {
    updateJob(childPaths, { delegate_pid: delegatePid });
    opts.updateParentJob({
      delegate_pid: delegatePid,
      current_step_job_path: childDir,
    });
  }

  function onEvent(
    event: Record<string, unknown>,
    eventCount: number,
  ): void {
    parentEventCount += 1;
    const timestamp = utcNow();
    updateJob(childPaths, {
      event_count: eventCount,
      last_event_at: timestamp,
      last_event_type: event.type,
    });
    opts.updateParentJob({
      event_count: parentEventCount,
      last_event_at: timestamp,
      last_event_type: event.type,
      current_step_job_path: childDir,
    });
  }

  let envelope: Record<string, unknown>;
  try {
    envelope = await executeRequestPipeline(childRequest, childDir, {
      onSpawn,
      onEvent,
    });
  } catch (exc) {
    writeFailureEnvelope(
      childRequest,
      childDir,
      "workflow_step_error",
      String(exc),
      { exitCode: 1 },
    );
    updateJob(childPaths, {
      state: "failed",
      finished_at: utcNow(),
      current_role: null,
      delegate_pid: null,
      last_error: String(exc),
    });
    opts.updateParentJob({
      delegate_pid: null,
      current_step_job_path: null,
    });
    throw exc;
  }

  updateJob(childPaths, {
    state: finalizeJobState(envelope),
    finished_at: utcNow(),
    assistant_role:
      (envelope.assistant_role as string) ??
      (childRequest.assistant_role as string),
    current_role: null,
    delegate_pid: null,
    last_error: envelope.error_message ?? null,
    task_type:
      (envelope.task_type as string) ??
      (childRequest.task_type as string) ??
      null,
    workflow_roles:
      (envelope.workflow_roles as string[]) ??
      (childRequest.workflow_roles as string[]),
    workflow_step: 1,
    workflow_total_steps: 1,
  });
  opts.updateParentJob({
    delegate_pid: null,
    current_step_job_path: null,
  });
  return [envelope, parentEventCount];
}

export async function executeWorkflow(
  request: Record<string, unknown>,
  artifactsDir: string,
  opts: {
    updateParentJob: (changes: Record<string, unknown>) => void;
  },
): Promise<Record<string, unknown>> {
  const paths = artifactPaths(artifactsDir);
  const workflowEnvelope = baseEnvelope(paths, request);
  workflowEnvelope.assistant_role = "supervisor";
  workflowEnvelope.task_type = (
    request.task_packet as Record<string, unknown>
  ).task_type;
  workflowEnvelope.workflow_roles = request.workflow_roles;
  workflowEnvelope.lineage = request.lineage ?? null;
  workflowEnvelope.completion_contract = {
    name: (request.completion_contract as Record<string, unknown>).name,
    role: (request.completion_contract as Record<string, unknown>).role,
  };

  const stepArtifactsDir = path.join(artifactsDir, "steps");
  mkdirSync(stepArtifactsDir, { recursive: true });
  (workflowEnvelope.artifacts as Record<string, unknown>).steps_dir =
    stepArtifactsDir;

  const priorSteps: Record<string, unknown>[] = [];
  const stepSummaries: Record<string, unknown>[] = [];
  const stepEnvelopes: Record<string, unknown>[] = [];
  let parentEventCount = 0;
  const workflowRoles = request.workflow_roles as string[];

  for (let index = 0; index < workflowRoles.length; index++) {
    const role = workflowRoles[index];
    const stepNum = index + 1;

    opts.updateParentJob({
      current_role: role,
      workflow_step: stepNum,
      workflow_total_steps: workflowRoles.length,
      completed_roles: stepSummaries.map(
        (s) => s.role as string,
      ),
    });

    const childDir = path.join(
      stepArtifactsDir,
      `${String(stepNum).padStart(2, "0")}-${role}`,
    );
    mkdirSync(childDir, { recursive: true });

    const childRequest = prepareRoleRequest(
      request,
      role,
      priorSteps,
      paths.artifactsDir,
    );

    const [childEnvelope, newParentEventCount] =
      await executeChildStep(childRequest, childDir, {
        role,
        workflowStep: stepNum,
        updateParentJob: opts.updateParentJob,
        parentPaths: paths,
        parentEventCount,
      });

    parentEventCount = newParentEventCount;
    stepEnvelopes.push(childEnvelope);
    const summary = stepSummary(childEnvelope, role, childDir);
    stepSummaries.push(summary);
    priorSteps.push(summary);

    if (role !== "critic" && !(childEnvelope.ok as boolean)) {
      break;
    }
  }

  const finalChild =
    stepEnvelopes.length > 0
      ? stepEnvelopes[stepEnvelopes.length - 1]
      : {};
  const allFindings: unknown[] = [];
  const criticFindings: unknown[] = [];
  const openRisks: string[] = [];
  let totalCostUsd = 0;
  let totalCostSeen = false;
  let durationMs = 0;
  let durationSeen = false;
  let numTurns = 0;
  let turnsSeen = false;
  const permissionDenials: Record<string, unknown>[] = [];
  const toolUses: Record<string, unknown>[] = [];
  let toolUseCount = 0;

  for (const summary of stepSummaries) {
    allFindings.push(
      ...((summary.findings as unknown[]) ?? []),
    );
    if (summary.role === "critic") {
      criticFindings.push(
        ...((summary.findings as unknown[]) ?? []),
      );
    }
    openRisks.push(
      ...((summary.open_risks as string[]) ?? []),
    );
  }

  for (const envelope of stepEnvelopes) {
    const cost = envelope.total_cost_usd;
    if (typeof cost === "number") {
      totalCostSeen = true;
      totalCostUsd += cost;
    }
    const childDuration = envelope.duration_ms;
    if (typeof childDuration === "number" && Number.isInteger(childDuration)) {
      durationSeen = true;
      durationMs += childDuration;
    }
    const childTurns = envelope.num_turns;
    if (typeof childTurns === "number" && Number.isInteger(childTurns)) {
      turnsSeen = true;
      numTurns += childTurns;
    }
    permissionDenials.push(
      ...((envelope.permission_denials as Record<string, unknown>[]) ?? []),
    );
    toolUses.push(
      ...((envelope.tool_uses as Record<string, unknown>[]) ?? []),
    );
    toolUseCount += Number(
      envelope.tool_use_count ??
        (envelope.tool_uses as unknown[])?.length ??
        0,
    );
  }

  const stepsFailed = stepEnvelopes.some(
    (envelope) => !(envelope.ok as boolean),
  );
  const boundary = aggregateBoundary(stepEnvelopes);
  const verification = aggregateVerification(stepEnvelopes);
  const status = stepsFailed
    ? "failed"
    : criticFindings.length > 0
      ? "needs_review"
      : "completed";

  workflowEnvelope.workflow = {
    roles: workflowRoles,
    steps: stepSummaries,
  };
  workflowEnvelope.structured_output = {
    status,
    summary:
      ((finalChild.completion as Record<string, unknown>) ?? {})
        .summary ??
      (finalChild.result as string) ??
      "workflow finished",
    steps: stepSummaries.map((item) => ({
      role: item.role,
      job_path: item.job_path,
      ok: item.ok,
      summary: item.summary,
    })),
    open_risks: dedupeStrings(openRisks),
  };
  workflowEnvelope.completion = workflowEnvelope.structured_output;
  workflowEnvelope.findings = allFindings;
  workflowEnvelope.open_risks = dedupeStrings(openRisks);
  workflowEnvelope.result = finalChild.result ?? null;
  workflowEnvelope.total_cost_usd = totalCostSeen
    ? totalCostUsd
    : null;
  workflowEnvelope.duration_ms = durationSeen ? durationMs : null;
  workflowEnvelope.num_turns = turnsSeen ? numTurns : null;
  workflowEnvelope.permission_denials = permissionDenials;
  workflowEnvelope.tool_use_count = toolUseCount;
  workflowEnvelope.tool_uses = toolUses;
  workflowEnvelope.boundary = boundary;
  workflowEnvelope.verification = verification;
  workflowEnvelope.session_id = request.session_id;
  workflowEnvelope.ok =
    !stepsFailed && criticFindings.length === 0;

  if (stepsFailed) {
    const failedStep =
      stepEnvelopes.find((e) => !(e.ok as boolean)) ?? {};
    workflowEnvelope.error_type =
      (failedStep.error_type as string) ?? "workflow_step_failed";
    workflowEnvelope.error_message =
      (failedStep.error_message as string) ?? "workflow step failed";
  } else if (criticFindings.length > 0) {
    workflowEnvelope.ok = false;
    workflowEnvelope.error_type = "critic_findings";
    workflowEnvelope.error_message = "critic found issues";
  }

  opts.updateParentJob({
    completed_roles: stepSummaries.map(
      (step) => step.role as string,
    ),
    current_role: null,
    workflow_step: stepSummaries.length,
  });

  writeJson(paths.normalized, workflowEnvelope);
  writeJson(paths.handoff, shapeHandoff(workflowEnvelope));
  return workflowEnvelope;
}
